### ISSUE TITLE
architect: learned routability factor for placement GA fitness function (Phase 0 — feasibility experiment)

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/kicad-tools#3185]
## Motivation

The placement GA's fitness function is the load-bearing component that decides whether the optimizer converges on a manufacturable layout. We've iterated on it repeatedly — PR #3114 added `--use-routing-fitness` (running the C++ A* router as the fitness oracle), PR #3120 added the `nets_fully_connected` lex-tuple metric. But each iteration is hand-engineered, and the genuinely-expensive ground-truth oracle (the A* router) is too slow to use in a tight GA loop: per the #3127 builder, `--use-routing-fitness` over 200 evaluations takes 3–6 hours on board 05.

The proposal: **train a small model to predict routability from placement features, and use its output as a weighted factor in the GA's FOM.** A learned predictor that runs in milliseconds can act as a cheap proxy for the expensive router oracle, letting the GA explore far more candidates per wall-clock minute.

This isn't an attempt to *replace* the existing FOM or the actual router — it's adding one weighted term whose job is to filter out obviously-unroutable placements early, while the existing congestion/length/symmetry terms still dominate.

## Phase 0 scope (this issue)

**A single, scoped feasibility experiment** that answers one question: *does any learnable signal exist between cheap placement features and actual routing outcomes on our boards?*

If the answer is yes (e.g., R² > 0.5 on holdout), Phase 1 (proper model + integration) gets its own issue. If no, we've learned that for ~1-2 days of work instead of months.

### Procedure
1. **Single-board corpus**: pick board 05 (the hardest of the in-repo boards; M-E milestone).
2. **Generate ~200 perturbations**: jitter component positions ±5mm, swap symmetric pairs, vary rotation by ±90°. Keep schematic connectivity fixed.
3. **Label each by running the actual router**: capture routing reach (nets fully connected / total), DRC blocking count, wall-clock. Use `--backend cpp` for speed.
4. **Hand-craft 10–20 features per placement**:
   - Pin density per quadrant + per dense-package neighborhood
   - Max channel width estimate (closest free corridor between obstacle clusters)
   - RSMT total length over signal nets
   - Component-to-edge minimum, pin-to-pin minimum
   - Dense-package count (pre-flag via `is_dense_package`)
   - Inter-component proximity for analog vs digital nets (reuse Phase 2's `detect_analog_nets`)
5. **Fit a gradient-boosted regressor** (`scikit-learn` or `xgboost`) against routing reach + DRC count.
6. **Report**: R² on held-out 30%, feature importances, calibration plot, qualitative analysis of where the model is wrong.

**Deliverable**: a writeup (markdown in `docs/research/`) documenting the experiment, results, and a recommendation: *escalate to Phase 1 (NN + GA integration)*, *iterate features*, or *abandon*.

## Investigation: external training corpora

@rwalters' observation: there may be existing labeled corpora out there. Worth a parallel sub-investigation:

- **CircuitNet** (ASIC, not PCB, but their congestion-prediction features may transfer)
- **DeepPCB** (defect detection dataset — wrong target, but the layout images may be useful for unsupervised pretraining)
- **OSHW / OpenHardwareCertified** project archives (have manufactured-confirmed labels; many are KiCad)
- **GitHub corpus**: `language:kicad_pcb` search returns thousands of files, but mostly unlabeled
- **JLCPCB / PCBWay public order archives**: if either publishes accepted-design metadata, that's near-ideal positive labels

The honest dataset bottleneck is *labels*, not placements. An external corpus of \"manufactured-confirmed\" PCBs gives us many positive examples without running the router, which reframes Phase 0 as a learning-to-rank or one-class classification rather than full regression. That weakens the data requirement substantially.

**Phase 0 deliverable for this thread**: a separate `docs/research/external_pcb_corpora_survey.md` enumerating what exists, license/format compatibility, and whether any would replace or augment the synthetic-perturbation pipeline.

## Architecture sketch (post-Phase 0)

```
┌─────────────┐     ┌──────────────────┐     ┌─────────────┐
│ Placement   │ ──> │ Feature extractor│ ──> │ Predictor   │ ──> routability score (0..1)
│ candidate   │     │ (hand-crafted    │     │ (MLP, ~1ms) │
└─────────────┘     │  + congestion    │     └─────────────┘
                    │  raster)         │
                    └──────────────────┘

GA fitness:
  fom = α·congestion + β·length + γ·symmetry + δ·learned_routability + ε·...
        (existing terms)                    ↑
                                            (NEW; small δ initially)
```

## Acceptance criteria (Phase 0 only)

1. Perturbation script + router-label runner committed under `scripts/research/`. Reproducible with one command.
2. Labeled dataset (≥150 successful router runs) committed as a small CSV / parquet under `data/research/` (or generated lazily on demand).
3. Feature-extractor code + baseline gradient-boost model committed.
4. Writeup at `docs/research/learned_fom_phase0.md` with R², feature importances, calibration plot, and recommendation.
5. External-corpus survey at `docs/research/external_pcb_corpora_survey.md`.
6. Issue closes with one of: \"recommend Phase 1\", \"recommend feature iteration\", or \"recommend abandon — predictive signal too weak\".

## Out of scope (deferred)

- Training a neural network (MLP/CNN/GNN). Phase 0 is gradient-boost only — NN comes in Phase 1 if and only if features-only baseline saturates.
- Integration into the actual GA fitness function. Phase 0 produces a predictor; Phase 1 weights it into FOM.
- Multi-board generalization. Phase 0 trains and tests on board 05 only; cross-board transfer is Phase 2.
- Production inference pipeline. Phase 0 is an offline notebook-style experiment.
- Active learning / RL placement (DREAMPlace-style end-to-end). Way out of scope.

## Risks and open questions

1. **Data scarcity / diversity**: 200 perturbations of one board mostly teach the model \"what board 05 placements look like.\" The model may not generalize even within board 05's design space. Mitigation: hold out perturbations from independent random seeds, not random shuffles.
2. **Calibration**: routing outcomes are heavy-tailed (most placements route poorly, a few route well). A poorly-calibrated predictor could overweight rare events. Use stratified sampling and check residual distribution.
3. **Feature-engineering ceiling**: if hand-crafted features cap at R² ~0.3, NNs may help — but they may also just overfit a 200-sample dataset. Don't escalate to NNs before checking that the feature space is genuinely too rich for trees.
4. **Confounder risk**: the model may learn \"more vias = bad\" because in our specific board this happens to correlate with poor routing. Inspect feature importances skeptically.
5. **Inference latency budget**: the predictor must run in << router time to pay back. A tree model is fine; a large NN may not be. Phase 0 doesn't hit this but Phase 1 must measure it.

## References

- PR #3114 — `--use-routing-fitness` (the slow oracle this would augment)
- PR #3120 — `nets_fully_connected` lex-tuple metric (current FOM primary key)
- PR #3127 — board-05 placement reflow that hit the 3-6h budget limit
- PR #3140 — partner-aware A* heuristic (recent FOM/heuristic improvement)
- DREAMPlace (NVIDIA, 2020) — learned ASIC placement
- CircuitNet (Tsinghua, 2022) — ML for ASIC EDA datasets

## Estimated effort (Phase 0)

- Perturbation + label runner: ~150 LOC + a few hours of compute
- Feature extractor: ~200 LOC
- Modeling notebook: ~100 LOC
- External corpus survey: ~1 day of investigation + writeup
- Total: roughly 1-2 weeks of focused work, mostly offline experiments

**Comment by rjwalters** [comment]
## External PCB corpus survey (2026-05-29)

Investigation of what publicly-accessible PCB layout corpora exist. TL;DR: **no ready-made labeled (placement, routing-outcome) dataset exists for PCBs**; what does exist is a usable starting point but still requires labeling by running our router. The most promising leads are #1 (kicad-happy-testharness) and #2 (OSHWA), with #3 (academic) being methodologically interesting but format-incompatible.

### Concrete corpora

#### 1. kicad-happy-testharness — `aklofas/kicad-happy-testharness`
- **Size**: 5,822 real-world KiCad projects, sourced from GitHub/GitLab/Codeberg via crawl (`search_repos.py`).
- **Format**: a master list (`repos.md`) of repo URLs + pinned commit hashes. The corpus itself is NOT redistributed — `checkout.py` clones each locally.
- **Metadata**: per-project KiCad file inventory + analyzer outputs (schematic, PCB, Gerber analyses) + complexity/quality scores in `repo_catalog.json`. **2.1M+ machine-checkable assertions across the corpus.**
- **License**: harness is open source (~MIT-ish), no explicit license statement on the corpus list itself. Individual project licenses vary; downstream redistribution would need per-project audit.
- **Manufacturability labels**: NONE explicit. \"Quality\" is flagged via assertions and findings, not manufactured-confirmed.
- **Usefulness for us**: very high — gives us a curated index of ~5.8K real KiCad projects without crawling ourselves. We'd need to run our router on each (or a stratified subset) to generate (placement, routing-outcome) labels.

#### 2. OSHWA Certified Projects Directory — `certification.oshwa.org`
- **Size**: 1,000+ projects (entries through US000325+ visible; ~80 countries, ~15 project types).
- **API**: public endpoint at `https://certificationapi.oshwa.org` (auth/rate-limit details need direct probing — docs page is JS-rendered, WebFetch didn't capture).
- **Metadata per project**: name, UID, type (Electronics is one of 15), creator, country, hardware license (CERN/TAPR/GPL/CC-BY-SA/etc.), software license, descriptive tags.
- **CAD format filtering**: NOT a top-level filter; individual project pages contain repo URLs that we'd need to scrape and inspect for `.kicad_pcb` content.
- **Manufacturability labels**: implicit — every OSHW-certified project has been documented for buildability. This is a *positive-label* corpus (\"these designs were good enough to be certified\") without negative examples.
- **Usefulness for us**: medium-high. Smaller than #1 but every entry is implicitly a positive example, which is exactly the regime where one-class classification or learning-to-rank weakens the data requirement. The KiCad subset is probably hundreds of projects, not thousands.

#### 3. Academic ASIC datasets — for methodology transfer only
- **CircuitNet** (Chai et al. 2022; CircuitNet-N14 2024) — 10K+ samples for VLSI tasks (congestion, DRC, IR drop, timing). Open source; new version on Hugging Face. **ASIC, not PCB** — LEF/DEF + congestion-heatmap labels.
- **ISPD/ICCAD contest benchmarks** (2011–2019) — LEF/DEF placement+routing benchmarks. **ASIC.**
- **Usefulness**: format-incompatible (LEF/DEF vs `.kicad_pcb`), but the **labeling methodology** (congestion-heatmap per-cell as the ML target) and the **feature engineering** (pin density rasters, placement images) transfer directly. CircuitNet's congestion-map target is the exact thing we'd want a placement-quality predictor to produce.

#### 4. DeepPCB — explicitly NOT what we want
- 1,500 PCB **image pairs** for trace/via defect detection (open/short/mousebite/spur/copper/pinhole). Computer-vision dataset, not layout-quality. Mentioned only to dismiss it from confusion.

#### 5. SparkFun / Adafruit official sources
- SparkFun maintains KiCad libraries (footprints/symbols), not full projects, on GitHub. Project source files were historically Eagle. Some converted; gradual KiCad migration as of 2025.
- Adafruit: Eagle-only officially. Conversion possible (`easyeda2kicad6`, `uPesy/easyeda2kicad.py`) but lossy.
- **Usefulness**: low for a layout corpus. Useful as footprint sources we already cover via KiCad's bundled libs.

#### 6. EasyEDA / LCSC — possibly large but no clean export
- EasyEDA hosts a large open-projects archive but no scrape-friendly export. Conversion tools exist (`easyeda2kicad6`) but project-level export is per-project.
- **Usefulness**: low-medium. Possible long-term data source if we build a conversion+harvest pipeline; not Phase 0 material.

### Methodological prior art (papers worth reading if Phase 0 proceeds)

- **Wei et al., \"CNN routability predictor\"** (cited in some 2025 reviews) — trained on **372 benchmarks → 26K placement images**. ASIC-focused but the (place-image → CNN → routability) recipe is our template. Suggests a ~70× perturbation factor per board to hit ~26K samples.
- **\"Painting on Placement\" (DAC 2019, Liu et al.)** — conditional GAN for routing congestion from placement images. **FPGA, not PCB**, but the image-to-image congestion-map formulation is what we'd consume.
- **PCB-Placement-GNN (June 2025)** — bipartite-graph PCB placement via GNN; the most directly-relevant recent prior art. Suggests GNN may outperform CNN for the netlist-aware portion (graph captures connectivity that rasters don't).
- **DREAMPlace** (NVIDIA) — learned ASIC placement; widely-cited but probably overkill for our Phase 0.

### Recommendation: Phase 0 bootstrap strategy

1. **Cheapest, in-house first**: as #3185 originally proposed, generate ~200 perturbations of board 05 + label by running our router. ~1 day of work, fully under our control, mirrors the in-repo boards.
2. **In parallel, pilot a 50-project slice from kicad-happy's `repos.md`** — pick projects with similar size/complexity to ours, run our router on each, see how many actually route end-to-end. This calibrates whether the broader corpus is usable.
3. **OSHWA API probe** — write a small script: hit `certificationapi.oshwa.org`, filter type=Electronics + license-compatible, scrape per-project repos to find KiCad files. This is cheap (a few hours) and tells us the actual positive-corpus size.
4. **Defer EasyEDA harvesting and academic-dataset conversion** — both are non-trivial pipelines that only pay off if Phase 0 shows learnable signal.

### Concrete corpus-pipeline issues to file post-#3185 approval

- **#3185-a (sub-task)**: enumerate kicad-happy `repos.md`, filter for projects routable by our pipeline, generate first-pass labels.
- **#3185-b (sub-task)**: probe OSHWA API, harvest KiCad-format certified projects into a positive-label corpus.
- **(Phase 1 follow-up)**: if Phase 0 succeeds, package the pipeline as `scripts/research/build_corpus.py` so the dataset grows monotonically.

### Key honest caveats

- **Label cost dominates**: regardless of corpus source, we still need (placement, routing-outcome) pairs. The router has to run for each. At ~few-minutes per board × thousands of projects, this is a real compute cost (hours-to-days, not weeks, but not free).
- **License audit needed before redistribution**: even with permissive per-project licenses, an aggregated labeled dataset published by us would need a per-project license check. For internal research use this isn't blocking.
- **Distribution shift**: a model trained on board-05-like designs (microcontroller + motor driver) may not predict well on RF or BGA-dense layouts. Phase 0 should explicitly hold out at least one board with a different topology to measure this.

**Comment by rjwalters** [comment]
## Closing as superseded by a refined three-issue decomposition (2026-05-29)

After conversation, the original Phase 0 framing has been refined significantly. Closing #3185 in favor of three focused issues that better match what we actually want to build.

### Refinement summary

**1. Binary labels, not regression.** A board ships iff it passes DRC + LVS + ERC + manufacturer-tolerance allowlist. That's a Boolean AND of hard constraints — exactly the {0,1} reward signal a PPO-style learned residual wants. Reframes the model from \"predict routing reach\" (continuous, fuzzy) to \"predict ship/no-ship\" (binary, calibrated). Changes the metric from R² to AUC + calibration.

**2. Hybrid FOM structure.** The figure of merit should be:
- **Hard constraints** as a goto-fail gate (DRC/LVS/ERC pass → continue, else FOM=0).
- **Soft objectives** as a log-linear weighted sum of normalized, hand-engineered terms.
- **Learned residual** as a multiplicative modifier `predictor(placement)^β`, where β=0 falls back cleanly to the hand-engineered FOM.

The hand-engineered FOM is independently valuable. Even without any ML, a properly-decomposed multi-term soft FOM is a substantial upgrade over the current `--use-routing-fitness` lever, ships immediately, and provides the baseline the learned residual has to beat.

**3. Corpus strategy clarified.** kicad-happy-testharness (5,822 projects) is positive-only and label-cost-bound (~5 min × thousands of projects to label). Practical Phase 0 uses 20 *diverse* seeds → 30 perturbations each = 600 labels, ~2 days of compute. Big enough to see signal, small enough to iterate.

**4. Specific soft-FOM terms enumerated** (see #NEW-A for the full table). Highlights:
- Trace length normalized by Steiner-minimum-tree lower bound (not absolute).
- Vias weighted by mfg cost class (micro vs standard vs blind/buried).
- Turning-angle penalty as mean-square deviation from 0°/45° grid.
- Net congestion variance across board cells (most directly correlated with routability).
- Decoupling proximity, thermal spread, crossing minimization.

### Successor issues

- **#NEW-A** (`loom:architect`): hybrid FOM with hard-constraint gate + multi-term soft objectives. Ships independent of ML. ~3-4 weeks.
- **#NEW-B** (`loom:architect`): learned residual Phase 0 — corpus + perturbation + gradient-boost classifier. Depends on NEW-A landing for FOM integration point. ~2 weeks.
- **#NEW-C** (`loom:architect`): FOM weight calibration — Pareto sweep across the 8 in-repo boards to fit soft-term weights. Depends on NEW-A. ~1 week.

External corpus survey from earlier in this thread is preserved as a comment on NEW-B.

**Comment by rjwalters** [comment]
Successor issues filed: **#3186** (hybrid FOM structure, no ML), **#3187** (learned-residual Phase 0 binary classifier), **#3188** (FOM weight calibration via Pareto sweep).
